clang/test/SemaSYCL/sycl-restrict.cpp

```diff
@@ -180,16 +180,15 @@ int use2 ( a_type ab, a_type *abp ) {
 
 template <typename name, typename Func>
 __attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
-  // expected-note@+1 {{called by 'kernel_single_task}}
   kernelFunc();
   a_type ab;
   a_type *p;
+  // expected-note@+1 {{called by 'kernel_single_task'}}
   use2(ab, p);
 }
 
 int main() {
   a_type ab;
-  // expected-note@+1 {{called by 'operator()'}}
```

Comment:
Actually this diagnostics was very useful, because it points to user's kernel which caused problems. Now call stack notes ends on functions with `sycl_kernel` attribute which are defined inside the SYCL headers, so if user has multiple kernel runs, notes won't help to find problematic one...

Reply:
It will only do that if the kernel is absolutely wrong (and hopefully we won't release a set of SYCL headers with a template that is impossible to use).  Consider:

    template<typename T> void foo(T t) {
        NonExistentFunctionCall();  // 1
        NonExistentFunctionCall(t); // 2
    }

In the above example, 1 can be diagnosed the very first time we look at 'foo', see https://godbolt.org/z/Kpx7AR .  However, 2 could possibly depend on the type of 'T' (it is so called, 'type dependent').  Thus we cannot see it.  Because we caught the error immediately (rather than later), we don't HAVE a callstack/instantiation stack to print.  It would give an error the very first time (and every time) someone #included this file, so we likely wouldn't let this get past testing.

However, look at the message for 2: https://godbolt.org/z/cxj97F .  It can only be issued once you try to instantiate the template. Thus, we can put the instantiation/call stack there(because now it exists!).

Reply:
Okay, thanks for the explanation.
Do you plan to add emitting diagnostics during instantiation too (I see that this patch is only *part* of https://github.com/intel/llvm/issues/1018 fix)?
I personally thought that note which points to problematic user's kernel run is a bit more useful than errors which are useful only for SYCL headers developers. @bader , please provide your opinion on this, If I'm wrong.

Reply:
I don't think we would want to diagnose 2x, right?  Getting 2 errors for this seems wrong.

The idea here is that if it is OUR headers, we wouldn't release them with this error happening (which we would catch before a user ever saw our header), but if it was something user caused, it would print the stack trace.

Reply:
According to SYCL language, user code here also is lambda passed to `kernel_single_task`. It calls `usage` which contains call to `eh_not_ok` function (so all of these functions is potential user's code) and it uses exceptions. As far as I understand we diagnose exceptions usage in device code through deferred diagnostics system, so, I'd expected call stack note here then.

Reply:
The way our delayed diagnostics work, we only diagnose each location once, even if it is called multiple times. We don't call it every time it is called.  Just like how the rest of diagnostics works, we only point out an error once (obviously with the exception of the bug we saw before).

We diagnose it being called through useage2 instead.

Reply:
Okay, understand.
Could you please add a test case which will show full call stack notes, to make sure that we won't broke it with the next changes?

Reply:
This test is showing the full notes, though they are quite disorganized because of all the diagnostics in this file. Do you want a new test file with them better organized? Or did you have something else in mind?

Reply:
I just wanted the test case with call stack notes pointing to all the functions starting from restriction usage and finishing by user's kernel run. Like this:
```
// let's have regular "sycl_kernel" function
template <typename name, typename Func>
void kernel_single_task(Func kernelFunc) {	
// expected-note@+1 {{called by 'kernel_single_task}}
  kernelFunc();
}

void foo() {
  // expected-error@+1 {{SYCL kernel cannot use exceptions}}
 throw 3;
}

...
// Somewhere inside the main:
// expected-note@+1 {{called by 'operator()'}}
kernel_single_task<class fake_kernel>([]() { some_foo(3); });
```